File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int solve(TreeNode root) {
        if(root == null) return 0;
        int lh = solve(root.left);
        if(lh == -1) return -1;
        int rh = solve(root.right);
        if(rh == -1) return -1;
        if(Math.abs(lh - rh) > 1) return -1;
        return Math.max(solve(root.left), solve(root.right)) + 1;
    }
    public boolean isBalanced(TreeNode root) {
        return solve(root) == -1 ? false : true;
    }
}

/*

class Solution {
    public int height(TreeNode root) {
        if(root == null) return 0;
        return 1 + Math.max(height(root.left), height(root.right));
    }
    public boolean solve(TreeNode root) {
        if(root == null) return true;
        int lh = height(root.left);
        int rh = height(root.right);
        if(Math.abs(lh - rh) > 1) return false;
        return solve(root.left) && solve(root.right);
    }
    public boolean isBalanced(TreeNode root) {
        return solve(root);
    }
}

*/

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root) {
        if(!root) return 0;
        int lh = solve(root->left);
        if(lh == -1) return -1;
        int rh = solve(root->right);
        if(rh == -1) return -1;
        if(abs(lh - rh) > 1) return -1;
        return max(lh, rh) + 1;
    }
    bool isBalanced(TreeNode* root) {
        return solve(root) == -1 ? false : true;
    }
};
/*
class Solution {
public:
    int height(TreeNode* root) {
        if(!root) return 0;
        return 1 + max(height(root->left), height(root->right));
    }
    bool solve(TreeNode* root) {
        if(!root) return true;
        int lh = height(root->left);
        int rh = height(root->right);
        if(abs(lh - rh) > 1) return false;
        return solve(root->left) && solve(root->right);
    }
    bool isBalanced(TreeNode* root) {
        return solve(root);
    }
};
*/